### PR TITLE
Remove automatic setting of binary media types header

### PR DESCRIPTION
**Pull request by pierskarsenbarg**
The `binaryMediaTypes` property of the 'Api' resource of the `RestAPI` component would get overriden because of a default applied to an amazon header: `"x-amazon-apigateway-binary-media-types": ["*/*"]`.

This PR gets rid of it and adds a regression test, as well as upgrade tests in order to make sure we don't replace any resources in the process.

See comment here: https://github.com/pulumi/pulumi-aws-apigateway/issues/118#issuecomment-1850035758

Fixes: #118

### CODE REVIEW

**Comment by danielrbradley** [comment]
Is this not going to break anyone who's currently expecting the binary media types to be set automatically?

**Comment by pierskarsenbarg** [comment]
No because we're setting it here by default: https://github.com/pulumi/pulumi-aws-apigateway/blob/main/provider/cmd/pulumi-resource-aws-apigateway/apigateway/api.ts#L661 But this one <- was being overwritten by the swagger body if you updated the routes array.

File: provider/pkg/e2e_test.go [hunk 1]
```diff
@@ -15,7 +15,41 @@ import (
 )
 
 func TestApi(t *testing.T) {
-	e2eTest("api").Run(t)
+	e2eTest("api", providertest.WithBaselineVersion("1.0.1")).Run(t)
+}
+
+func TestBinaryMediaTypesNoInput(t *testing.T) {
+	e2eTest("binaryMediaTypesNoInput", providertest.WithProviderName("aws-apigateway"),
```
Comment (t0yv0) [inline provider/pkg/e2e_test.go:22]: Ah this is interesting, looks like this is included upgrade tests in this PR although I suspect there's no particular concern about testing upgrades for this new feature, that's just overhead. CC @danielrbradley FYI per our recent conversation, it might be less confusing to separate out upgrade tests into their own `func Test*` roots.

File: provider/pkg/e2e_test.go [hunk 2]
```diff
@@ -15,7 +15,41 @@ import (
...
+		providertest.WithSkippedUpgradeTestMode(
+			providertest.UpgradeTestMode_Quick,
+			"Quick mode is only supported for providers written in Go at the moment")).Run(t)
+}
+
+func TestSimpleNoBinaryMediaType(t *testing.T) {
+	e2eTest("simple").Run(t)
+}
+
+func TestSimpleWithBinaryMediaType(t *testing.T) {
```
Comment (t0yv0) [inline provider/pkg/e2e_test.go:33]: Do these tests fail before the changes to the provider? Or do we need to find a way to write some asserts?

File: provider/pkg/e2e_test.go (hunk 1 above)
Comment (VenelinMartinov) [reply provider/pkg/e2e_test.go:22]: We specifically wanted to add upgrade tests since the first iteration of this change forced replaces in resources.

@pierskarsenbarg  came up with a solution after which doesn't require that but I we need to add some tests for the various configurations of this, in order to make sure that none of them are broken by the changes.

File: provider/pkg/e2e_test.go (hunk 2 above)
Comment (VenelinMartinov) [reply provider/pkg/e2e_test.go:33]: Yeah, good point we should probably write tests for the actual bug fixed here.

File: provider/pkg/e2e_test.go (hunk 1 above)
Comment (t0yv0) [reply provider/pkg/e2e_test.go:22]: Ah excellent. So that's intentional then. Thank you!

File: provider/pkg/e2e_test.go (hunk 2 above)
Comment (t0yv0) [reply provider/pkg/e2e_test.go:33]: Yeah that would be ideal. If the tests still pass without the fix then if we silently regress we will not notice.

File: provider/pkg/test-programs/simple/index.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
+import * as apigateway from "@pulumi/aws-apigateway";
+import * as aws from "@pulumi/aws";
+
+import * as pulumi from "@pulumi/pulumi";
+
+const config = new pulumi.Config()
+
+const useSwaggerSpec = config.getBoolean("useSwaggerSpec") || true;
+const useBinaryMediaType = config.getBoolean("useBinaryMediaType") || true;
```
Comment (VenelinMartinov) [inline provider/pkg/test-programs/simple/index.ts:9]: Defaulting these to true must mean that we never actually test the false cases. Can we change these to false?

**Comment by VenelinMartinov** [comment]
Yup, I verified it here: https://github.com/pulumi/pulumi-aws-apigateway/actions/runs/7247196392/job/19740649953 and then restored the fix in the next commit.

**Review by thomas11 — APPROVED** [review]

**Comment by VenelinMartinov** [comment]
Note that users who previously specified `binaryMediaTypes` would get replacements after upgrading to this change.

This is probably fine since the resource wouldn't have worked correctly for them before and the resources are stateless.
